### ISSUE TITLE
Wish: Lock down paths manager

### ISSUE THREAD

**Issue by andydandy74** [DynamoDS/Dynamo#7274]
## Dynamo version

Any

## Operating system

Any

## Feature Request

When deploying Dynamo, I want to be able to lock down the Manage Paths dialog, so users cannot add new package paths. (In our Dynamo deployment, we include a customized version of DynamoSettings.XML that contains a central path to our curated set of packages and a local user-specific path for users to store their own stuff. What we don't want is for them to be able to add more paths in locations that may not be backupped.)

**Comment by sharadkjaiswal** [comment]
I am not sure how Dynamo behaves when DynamoSettings.xml is read-only. Ideally, all user settings could be frozen when this XML is read-only.

**Comment by andydandy74** [comment]
Even if that were possible users would lose access to their recent files on the start screen.

**Comment by radumg** [comment]
I second Andy's proposal and there's good precedents for this, for example Chrome displays a notice next to each setting that is managed by the deployment/enterprise.

Perhaps this could be a more general feature that lets us lock down *any* setting using the DynamoSettings.XML when doing deployments ?

**Comment by dbaldacchino** [comment]
If settings remain in a user .xml file though, it's up for grabs to edit. These settings should probably reside in the HKLM registry hive. Currently I have a login script that checks for the path in our library and adds it if not found so if the user takes it out, it comes back at next logon.

As for @andydandy74 's request, currently one could take out everything in the <CustomPackageFolders> node and rebuild it at each logon with the approved package paths only.

**Comment by andydandy74** [comment]
Surely possible, but I'd prefer a solution that doesn't add extra work for our IT staff.

**Comment by dbaldacchino** [comment]
Absolutely, I'm having to pick up the pieces or it doesn't get done :)

**Comment by radumg** [comment]
I know that feeling @dbaldacchino ! 

Yes, it would be up for grabs in XML, but we'd overwrite it at logon.
Also, if they're clever enough to mess with the XML and know where to look for it in the first place - they're most likely the ones that wouldn't do it. Just about anything can be bypassed nowadays, so we can't account for every possible hack.

**Comment by dbaldacchino** [comment]
Absolutely, but certain things should be easier to control. Those are things that IMO should be in settings locations that users cannot edit: either the HKLM hive or C:\ProgramData\etc.. It would avoid a lot of scripted hacks at logon etc.

**Comment by Racel** [comment]
@andydandy74 @dbaldacchino @radumg - What if we just provide a setting in the XML that will allow you to disable the Manage Paths dialog completely (prevent the dialog from even appearing). The XML file won't have to be read-only, but users can not easily use the UI to make changes until the BIM Manager enables it through the XML. Thoughts?

**Comment by radumg** [comment]
That would work too but @andydandy74 's suggestion of making the dialog read-only would be better as it would indicate to the users where these things are located. If they see our Design Technology/BIM/etc drive there for example, they'd know to come talk to the right people. 
More importantly though, from a UX point of view, it would indicate to users why that command isn't performing as expected. Chrome's precedent applies here again, a notice on the form would provide feedback why it's read-only. Simply not showing the form would make them think it isn't working (how many Github issues about this will people submit ?).

The menu item would have to greyed out too at the minimum if the form is disabled. 

Hopefully that makes sense :)

**Comment by Racel** [comment]
Thanks @radumg - I see what you mean. From a dev standpoint, it would probably be easier to just turn the dialog off completely (so there will be minimal UI work) and get this request in sooner than later. "Manage Paths..." has only been in Dynamo since 0.9.0. Like before, users will still have access to the Package Manger UI if they want to learn where packages are stored. 

The only downside is that BIM Managers will need to inform users where to find files and where to put their nodes if they want them to show up automatically. 

Keep in mind that turning the dialog off completely will only occur when someone manipulates the XML file intentionally.

**Comment by radumg** [comment]
thanks @Racel  - With regards to that downside, it's very likely that people won't have write access to the central repository of the nodes/packages library in a practice environment, so they'd have to speak to the managers anyway if they want their nodes to be included in the library. This is mostly for QA purposes. :+1: 

For example, our network location for nodes & packages libraries is readable/viewable by anyone but only users in a certain security group have write/modify permissions. (BIM Managers, Computational specialists, etc).

**Comment by andydandy74** [comment]
Same here. Packages are centrally curated (and vetted).

**Comment by dbaldacchino** [comment]
Same here, centrally located. However there should still be user settings available in the sake of nimbleness: if a user wants to try out a package, they should be able to install locally to test. Sounds like if you disable the dialog, completely, they wouldn't be able to do the same. At that point I'd rather retain flexibility and continue with my current strategy of having a part of the login script check that the central path is in the user .xml settings file and if not, add it.

**Comment by radumg** [comment]
To get around that issue @dbaldacchino mentioned of allowing people to still install packages, what we're doing currently is at logon time
- we copy over packages from our central location to their machines to standard Dynamo location
- overwriting any files that already exits
- but not deleting things that are on user machine & not in our central folder.

This makes sure all users have those approved packages from the central location, all on the same version, but still allows them to install additional packages locally.

The XML setting would allow us to prevent people from changing the default Dynamo location for example.
